src/tiled/objecttemplatemodel.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * objecttemplatemodel.cpp
+ * Copyright 2017, Thorbjørn Lindeijer <thorbjorn@lindeijer.nl>
+ * Copyright 2017, Mohamed Thabet <thabetx@gmail.com>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "objecttemplatemodel.h"
+
+#include "templategroup.h"
+
+namespace Tiled {
+namespace Internal {
+
+ObjectTemplateModel::ObjectTemplateModel(const TemplateDocuments &templateDocuments, QObject *parent):
+    QAbstractItemModel(parent),
+    mTemplateDocuments(templateDocuments)
+{
+}
+
+QModelIndex ObjectTemplateModel::index(int row, int column,
+                                       const QModelIndex &parent) const
+{
+    if (!parent.isValid()) {
+        if (row < mTemplateDocuments.size()) {
+            return  createIndex(row, column, mTemplateDocuments.at(row)->templateGroup());
+        } else {
+            return QModelIndex();
+        }
+    }
+
+    if (TemplateGroup *templateGroup = toTemplateGroup(parent)) {
+        if (row < templateGroup->templateCount())
+            return createIndex(row, column, templateGroup->templateAt(row));
+    }
+
+    return QModelIndex();
+}
+
+QModelIndex ObjectTemplateModel::parent(const QModelIndex &index) const
+{
+    if (!index.isValid())
+        return QModelIndex();
+
+    if (ObjectTemplate *objectTemplate = toObjectTemplate(index)) {
+        auto templateGroup = objectTemplate->templateGroup();
+        // TODO: this doesn't work properly when multiple documents have the same group
```

Comment:
This can't happen. You're comparing pointers here and at no point in the code would it insert the same pointer twice.